src/livox_gpu_points_plugin.cpp

```diff
@@ -0,0 +1,114 @@
+  GpuRayPlugin::Load(_parent, _sdf);
+  gazebo_node_ = transport::NodePtr(new transport::Node());
+  gazebo_node_->Init();
+
+  auto curr_scan_topic = _sdf->Get<std::string>("ros_topic");
+
+  // Create Node Handle
+  nh_ = new ros::NodeHandle();
+  pub_ = nh_->advertise<sensor_msgs::PointCloud2>(curr_scan_topic, 5);
+
+  raySensor = std::dynamic_pointer_cast<sensors::GpuRaySensor>(_parent);
+
+  raySensor->SetActive(true);
+  sub_ = gazebo_node_->Subscribe(raySensor->Topic(), &LivoxGpuPointsPlugin::OnNewLaserAnglesScans, this);
+}
+
+void
+LivoxGpuPointsPlugin::OnNewLaserAnglesScans(ConstLaserScanAnglesStampedPtr& _msg)
+{
+  const double MIN_RANGE = raySensor->RangeMin();;
+  const double MAX_RANGE = raySensor->RangeMax();;
+  const double MIN_INTENSITY = 0.0;
+
+  const unsigned int sampleSize = raySensor->SampleSize();
+
+  // Populate message fields
+  const unsigned int POINT_STEP = 22;
+  sensor_msgs::PointCloud2 msg;
+  msg.header.frame_id = raySensor->Name(); // laser_livox
+  msg.header.stamp = ros::Time(_msg->time().sec(), _msg->time().nsec());
+  msg.fields.resize(6);
+  msg.fields[0].name = "x";
+  msg.fields[0].offset = 0;
+  msg.fields[0].datatype = sensor_msgs::PointField::FLOAT32;
+  msg.fields[0].count = 1;
+  msg.fields[1].name = "y";
+  msg.fields[1].offset = 4;
+  msg.fields[1].datatype = sensor_msgs::PointField::FLOAT32;
+  msg.fields[1].count = 1;
+  msg.fields[2].name = "z";
+  msg.fields[2].offset = 8;
+  msg.fields[2].datatype = sensor_msgs::PointField::FLOAT32;
+  msg.fields[2].count = 1;
+  msg.fields[3].name = "intensity";
+  msg.fields[3].offset = 12;
+  msg.fields[3].datatype = sensor_msgs::PointField::FLOAT32;
+  msg.fields[3].count = 1;
+  msg.fields[4].name = "ring";
+  msg.fields[4].offset = 16;
+  msg.fields[4].datatype = sensor_msgs::PointField::UINT16;
+  msg.fields[4].count = 1;
+  msg.fields[5].name = "time";
+  msg.fields[5].offset = 18;
+  msg.fields[5].datatype = sensor_msgs::PointField::FLOAT32;
+  msg.fields[5].count = 1;
+  msg.data.resize(sampleSize * POINT_STEP);
+
+  uint8_t* ptr = msg.data.data();
+  for (unsigned int i = 0; i < sampleSize; i++)
+  {
+    double r = _msg->scan().ranges(i);
+    double intensity = _msg->scan().intensities(i);
+
+    // Get angles of ray to get xyz for point
+    double yAngle = _msg->scan().azimuth(i);
+    double pAngle = _msg->scan().zenith(i);
+
+    // pAngle is rotated by yAngle:
+    if ((MIN_RANGE < r) && (r < MAX_RANGE))
+    {
+      // ROS_INFO_STREAM("Range: " << r);
+      *((float*)(ptr + 0)) = r * cos(pAngle) * cos(yAngle); // x
+      *((float*)(ptr + 4)) = r * cos(pAngle) * sin(yAngle); // y
+      *((float*)(ptr + 8)) = r * sin(pAngle); // z
+      *((float*)(ptr + 12)) = intensity; // intensity
+      *((uint16_t*)(ptr + 16)) = 0; // ring
+      *((float*)(ptr + 18)) = 0.0; // time
+      ptr += POINT_STEP;
+    }
+  }
+  // Populate message with number of valid points
+  msg.data.resize(ptr - msg.data.data()); // Shrink to actual size
```

Comment:
I think this could be done nicer with `sensor_msgs::PointCloud2Modifier` and `Iterator` (compare e.g. [here](https://github.com/enwaytech/cleansquare_ros/blob/5a2a2b14a13d9cdbe4c638a7513ec5b44259f8b5/safety/blitz_one/collision_predictor/test/rostest/nodelet_tests.cpp#L258-L273)).